File: app/src/main/kotlin/com/mobileappdev/homeworkplanner/ClassSchedule.java
package com.mobileappdev.homeworkplanner;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.google.firebase.firestore.QuerySnapshot;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import static androidx.constraintlayout.widget.Constraints.TAG;

public class ClassSchedule {
    private static ClassSchedule sClassSchedule;
    private String uid = "";
    private FirebaseFirestore db = FirebaseFirestore.getInstance();

    private List<Class> mClasses;

    public static ClassSchedule get(Context context){
        if (sClassSchedule == null){
            sClassSchedule = new ClassSchedule(context);
        }
        return sClassSchedule;
    }

    private void createClass(Map<String, Object> dataMap) {
        Class aClass = new Class();
        aClass.setDaysOfWeek((ArrayList<String>) dataMap.get("dayOfWeek"));
        aClass.setClassName((String) dataMap.get("className"));
        aClass.setStartTime((String) dataMap.get("classStartTime"));
        aClass.setEndTime((String) dataMap.get("classEndTime"));
        aClass.setCreditHours((int) dataMap.get("creditHours"));
        mClasses.add(aClass);
    }

    private ClassSchedule(Context context){
        mClasses = new ArrayList<>();

        // TODO: Populate with classes from database
        if(FirebaseAuth.getInstance().getCurrentUser() != null){
            uid = FirebaseAuth.getInstance().getCurrentUser().getUid();
        }
        db.collection("user").document(uid).collection("classes")
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<QuerySnapshot> task) {
                        if (task.isSuccessful()) {
                            for (QueryDocumentSnapshot document : task.getResult()) {
                                Log.d(TAG, document.getId() + " => " + document.getData());
                                createClass(document.getData());
                            }
                        } else {
                            Log.d(TAG, "Error getting documents: ", task.getException());
                        }
                    }
                });
    }

    public void addClass(String className, String startTime, String endTime, ArrayList<String> daysWeek,
                    int creditHours)
    {
        Class aClass = new Class();
        aClass.setClassName(className);
        aClass.setStartTime(startTime);
        aClass.setEndTime(endTime);
        aClass.setDaysOfWeek(daysWeek);
    }

    public List<Class> getClasses(){ return mClasses; }
}

File: app/src/main/kotlin/com/mobileappdev/homeworkplanner/ClassSchedule.kt
package com.mobileappdev.homeworkplanner

import android.util.Log
import androidx.annotation.NonNull
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import java.util.ArrayList
import androidx.constraintlayout.widget.Constraints.TAG
import com.google.android.gms.tasks.Tasks
import com.google.firebase.auth.FirebaseUser
import kotlinx.coroutines.*

object ClassSchedule {
    private var uid = ""

    var mClasses: ArrayList<Class> = ArrayList()

    fun createClass(dataMap: Map<String, Any>) {
        val aClass = Class()
        aClass.daysOfWeek = dataMap["dayOfWeek"] as ArrayList<String>?
        aClass.className = dataMap["className"] as String?
        aClass.startTime = dataMap["classStartTime"] as String?
        aClass.endTime = dataMap["classEndTime"] as String?
        aClass.creditHours = dataMap["creditHours"] as Long
        mClasses.add(aClass)
    }

    fun initRoutine() = runBlocking {
        val db = FirebaseFirestore.getInstance()

        if (FirebaseAuth.getInstance().currentUser != null) {
            uid = FirebaseAuth.getInstance().currentUser!!.uid
            Log.d("uid:  ", uid)
        }

        mClasses.clear()
        //db.collection("user").document(uid).
        Tasks.await(db.collection("user").document(uid).collection("classes")
                .get()
                .addOnCompleteListener { task ->
                    if (task.isSuccessful) {
                        for (document in task.result!!) {
                            document.reference.collection("assignments")
                                    .get()
                                    .addOnCompleteListener { task2 ->
                                        for (assignment in task2.result!!) {
                                            val dataMap = assignment.data
                                            dataMap!!["className"] = document.data["className"]
                                            AssignmentList.createAssignment(dataMap)
                                        }
                                    }
                            Log.d(TAG, document.id + " => " + document.data)
                            createClass(document.data)
                        }
                    } else {
                        Log.d(TAG, "Error getting documents: ", task.exception)
                    }
                })
    }
}
